Frameworks/Foundation/_NSUndoGroup.mm

```diff
@@ -0,0 +1,114 @@
+//******************************************************************************
+//
+// Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <Foundation/NSUndoManager.h>
+#import "_NSUndoGroup.h"
+#import "_NSUndoCall.h"
+
+@implementation _NSUndoGroup : _NSUndoObject
+
+- (id)initWithLevel:(NSUInteger)level {
+    if (self = [super init]) {
+        _undoGrouping = [NSMutableArray array];
+        _undoLevel = level;
+        _isClosed = NO;
+    }
+    return self;
+}
+
+- (id)initWithOwner:(NSUndoManager*)manager {
+    if (self = [self initWithLevel:0]) {
+        _owningManager = manager;
+    }
+    return self;
+}
+
+- (void)addUndoCallToUndoGroup:(_NSUndoCall*)undoCall {
+    _NSUndoObject* topObjectInGroup = [_undoGrouping firstObject];
+    if ([topObjectInGroup isKindOfClass:[_NSUndoGroup class]] && ![topObjectInGroup isClosed]) {
+        [topObjectInGroup addUndoCallToUndoGroup:undoCall];
+    }
+    [_undoGrouping insertObject:undoCall atIndex:0];
+}
+
+- (void)callUndoGroup {
+    while ([_undoGrouping count] > 0) {
+        _NSUndoObject* object = [_undoGrouping firstObject];
+        if ([object isKindOfClass:[_NSUndoGroup class]]) {
+            [_owningManager callNestedGroup];
```

Comment:
If you make all of these conform to an internal interface such as `_NSUndoing`, you don't need to switch on the class.

`_NSUndoGroup`'s implementation of `_NSUndoing` would just call the `_NSUndoing` implementation on every one of its subjugates.

With Objective-C, class checking is a code smell.